% Define the module top matter
% This gets used to create the chapter title page
% NOTES:
%  * When multiple people have authored or contributed to the module, simply use a LaTeX line break
%    (a double-backslash: \\) at the end of each person.
%  * If you don't want this information shown on the module chapter page, simply remove the lines
%    within the \setModuleAuthors{} and \setModuleContributions{} environments
\setModuleTitle{Introduction to Snakemake}
\setModuleAuthors{%
  Nathan S. Watson-Haigh \mailto{nathan.watson-haigh@adelaide.edu.au}
}
%\setModuleContributions{%
%  John Doe \mailto{john.dow@example.com}%
%}

% BEGIN: Module Title Page
% This simply uses the above information and creates a module chapter page
% NOTES:
%  * The chapter page will always appear on odd numbered page
\chapter{\moduleTitle}
\newpage
% END: Module Title Page


% BEGIN: KLOs
% Key Learning Outcomes (KLOs) are an important aspect of any learning/training. They provide
% valuable infomation about what the trainee will have learned, what they will be able to do or know
% abouti at the end of the module. Unlike objectives which are more trainer oriented, KOLs are
% focused on the learner.
% At the end of the module, the KLOs can be used to develop criteria for writing an assessment to
% see if the trainees knowledge/skills have improved as a result of the module.
% 
% Search online for information on how to write KLOs. e.g.
% http://www.teaching-learning.utas.edu.au/__data/assets/word_doc/0014/23333/Learning-outcomes-v9.1.doc
\section{Key Learning Outcomes}

After completing this module the trainee should be able to:
\begin{itemize}
  \item Install Snakemake in a conda environment
  \item Execute a Snakemake workflow
  \item Use the provided "profile" to execute jobs on a compute cluster
  \item Write simple Snakemake rules capable of generating some output(s) by executing some code which perates on some input(s)
\end{itemize}
% END KLOs

% BEGIN: Resources Used
% This section can be used to describe the tools and data used during the module. It helps to act as
% a future reference to the trainee
\section{Resources Required}

For the purpose of this training you need access to:

\begin{itemize}
  \item A compute cluster with the \texttt{module} command available to you for loading software
  \item \url{https://sylabs.io/singularity/}{Singularity} - available as a module on the above cluster
  \item \url{https://www.anaconda.com/distribution/}{conda} - available as a module on the above cluster
\end{itemize}


\subsection{Tools Used}
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
  \item[Snakemake]\hfill\\
    \url{https://snakemake.readthedocs.io}
  \item[Graphviz]\hfill\\
    \url{https://www.graphviz.org}
\end{description}

\section{Useful Links}
 
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=0.5cm]
  \item[Slurm Documentation]\hfill\\
    \url{https://slurm.schedmd.com/documentation.html}
\end{description}

\newpage
% END: Resources Used

% BEGIN: Introduction
\section{Setting Up Your Environment}

For the purpose of the workshop we will be working on the head node of an HPC cluster running \url{https://slurm.schedmd.com/documentation.html}{Slurm}.
This is the most likely infrastructure that fellow bioinformaticians already find themselves using
on a regular basis. We also assume that the cluster provides the \texttt{module} command for you to
load software and the modules \texttt{Anaconda3} and \texttt{Singularity} are available to use.

The execution of the Snakemake workflow will actually take place on the cluster head node with jobs
being submitted to Slurm for queing and processing. From the head node, Snakemake will monitor the
submitted jobs for their completion status and submit new jobs as dependent jobs complete sucessfully.

\subsection{Connect to the Cluster Head Node}

\begin{steps}
First up, lets connect to the head node of the HPC cluster using \texttt{ssh}.

\emph{See your local facilitator for connection details. You should have one user account per person.}
\end{steps}

\subsection{Install Snakemake}

The recommended installation route for Snakemake is through a conda environemnt
(\url{https://snakemake.readthedocs.io/en/stable/getting_started/installation.html}). As such, you need
Anaconda3, usually avaiable to you on your cluster via the module system.

\begin{steps}
\begin{lstlisting}
# Load the Anaconda3 module on your cluster
# If it's unavailable contact the cluster sysadmin
module load \
  Anaconda3

# Create a new conda environment for snakemake
# We use a specific version for reproducibility reasons
# https://anaconda.org/search?q=snakemake
SNAKEMAKE_VERSION="5.5.4"

conda create \
  --name snakemake \
  --channel bioconda --channel conda-forge \
  --yes \
  snakemake="${SNAKEMAKE_VERSION}"
\end{lstlisting}

That's all there is to installing Snakemake.

To activate the environment and make \texttt{snakemake} available on the command line simply:

\begin{lstlisting}
# activate the conda environment you created above
source activate snakemake
\end{lstlisting}
\end{steps}


\begin{bonus}
While waiting for others to catch up, why not have a look into how you would go about updating \texttt{Snakemake}
within this conda environment if there is a new version available.

\begin{answer}
\begin{lstlisting}
conda update \
  --channel bioconda --channel conda-forge \
  snakemake
\end{lstlisting}
\end{answer}

\end{bonus}


\subsection{Running a Workflow}

We are going to work with an example Snakemake workflow to demonstrate how they are run before
we get stuck into writing our own. This example workflow consists of the following steps:

\begin{itemize}
  \item Running FastQC across the raw reads
  \item Aggregating the raw read FastQC reports using MultiQC
  \item Performing adapter, quality, and read length filtering using Trimmomatic
  \item Running FastQC across the QC'd reads
  \item Aggregating the QC read FastQC reports using MultiQC
  \item Index the reference FASTA file
  \item Perform a \texttt{bwa-mem} read alignment
\end{itemize}

\begin{steps}

Lets get the example workflow:

\begin{lstlisting}
# Clone the repository
git clone https://github.com/UofABioinformaticsHub/2019_EMBL-ABR_Snakemake_webinar ./example_workflow

cd example_workflow/snakemake-tutorial
\end{lstlisting}

The most basic of Snakemake commands will execute the workflow, running jobs on the head node (your cluster
sysadmin will probably growl at you). Lets be nice, and just do a \texttt{dryrun} for now:

\begin{lstlisting}
snakemake \
  --dryrun
\end{lstlisting}
\end{steps}

\begin{questions}

Using the Snakemake help, how do you get Snakemake to print the shell commands it will execute for each job?

\begin{answer}
\begin{lstlisting}
snakemake \
  --dryrun \
  --printshellcmds
\end{lstlisting}
\end{questions}














% To make a paragraph appear as a "note" to the reader, simply wrap it in a "note" environment like
% this:
\begin{note}
This is an important note to the reader. The paragraph get's it own margin icon and formatting to
ensure it stands out.
\end{note}

There are 10 characters which have special meaning in \LaTeX and to have them displayed as literal
characters we need to do something special. The first seven can simply be prepended by a backslash;
for the other three, we must use the macros \verb+\textasciitilde+, \verb+\textasciicircum+,
\verb+\textbackslash+.


\& \% \$ \# \_ \{ \}

\textasciitilde

\textasciicircum

\textbackslash

Highly redundant coverage ($>$15X) of the genome can be used to correct sequencing
errors in the reads before assembly and errors. Various k-mer based error
correction methods exist but are beyond the scope of this tutorial.

In order to use a single character to encode Phred qualities, ASCII characters
are used (\url{http://shop.alterlinks.com/ascii-table/ascii-table-us.php}). All ASCII characters have a decimal
number associated with them but the first 32 characters are non-printable (e.g.

Because ASCII characters $<$ 33 are non-printable, using the Phred+33 encoding was
not possible. Therefore, they simply moved the offset from 33 to 64 thus

schema it is not always possible to identify what encoding is in use. For
example, if the only characters seen in the quality string are (\texttt{@ABCDEFGHI}),
then it is impossible to know if you have really good Phred+33 encoded qualities
or really bad Phred+64 encoded qualities.

\subsection{Tables}

Writing tables in \LaTeX is HARD! Rather than write them from scratch, I'd head over to
\url{http://www.tablesgenerator.com/latex_tables}, generate the table you want using the interactive
page and copy the resulting \LaTeX into your source file.

\begin{tabular}{llr}
\hline
\multicolumn{2}{c}{Item} \\
\cline{1-2}
Animal    & Description & Price (\$) \\
\hline
Gnat      & per gram    & 13.65      \\
          & each        & 0.01       \\
Gnu       & stuffed     & 92.50      \\
Emu       & stuffed     & 33.33      \\
Armadillo & frozen      & 8.99       \\
\hline
\end{tabular}

\subsection{Figures}

No handout is complete without figures, screenshots or similar.

\begin{figure}[H]
\centering
\includegraphics[width=0.8\textwidth]{handout/bad_example.png}
\caption{Per base sequence quality plot for \texttt{bad\_example.fastq}.}
\label{fig:bad_example_untrimmed_plot}
\end{figure}

\subsection{Maths Environment}

I won't pretend to be an expert in \LaTeX but I do know that one of the strengths of \LaTeX is it's
ability to allow mathematical formulas to be constructed and typeset very easily. Firstly, you have
to tell \LaTeX what text is going to contain mathematical symbols.

For inline display, simply surround the relevant text by dollar signs: \verb+$Q(A) =-10 log10(P(\sim A))$+

For more complex mathematical formulas which need to be displayed separately, use the
\verb+\begin{displaymath} ... \end{displaymath}+ environment to wrap around your formulas.

\begin{displaymath}
Q(A) =-10 log10(P(\sim A))
\end{displaymath} 


\subsection{Questions and Answers}

It is useful to maintain the answers to questions together in the same \LaTeX source. By using the
\verb+\begin{answer} ... \end{answer}+ environment we can have the enclosed answers excluded from
the trainee's version of the handout while including it in the trainer's version of the handout.

\begin{questions}
We can ask a simply question to test if the trainee is understanding what they are doing. 
\begin{answer}
We can maintain the answer along side the questions.
\end{answer}

\end{questions}

\subsection{Code For Trainees to Copy/Type}

Different people have different opinions on whether it is a good idea to provide the commands for
trainees to copy-and-paste. In our experience, there is a huge amount of time wasted by novices
typing commands incorrectly or changing filenames which affects commands you might run later on. We
also think that it's a good idea to pose regular questions or ask the trainees to modify a previous
command. This way you can catch out those whoe are just trying to get ahead by blindly
copying-and-pasting.

To provide a nicely formatted, copy-and-pastable command, simply wrap it in the
\verb+\begin{lstlisting} ... \end{lstlisting}+ environment. Long commands will be wrapped
automatically and bash line continuation characters (\textbackslash) inserted where required.

\begin{steps}
\begin{lstlisting}
cd ~/QC
fastx_clipper -h
fastx_clipper -v -Q 33 -l 20 -M 15 -a GATCGGAAGAGCGGTTCAGCAGGAATGCCGAG -i bad_example.fastq -o bad_example_clipped.fastq
\end{lstlisting}
\end{steps}

